File: app-ios-nativa/Fontaneria/Pruebas/PruebasFontaneria/PruebasClienteApi.swift
import Dominio
import Foundation
import Testing

@testable import Fontaneria

/// En serie: todas comparten el mismo servidor de mentira, y en paralelo se
/// pisarían las respuestas unas a otras.
@Suite("Cliente del servidor")
struct PruebasClienteApi {
    private let cliente: ClienteApi
    private let buzon: ServidorFalso.Buzon

    init() {
        let (sesionHttp, buzon) = ServidorFalso.sesion()
        self.buzon = buzon
        cliente = ClienteApi(urlBase: "https://api.ejemplo.com", sesionHttp: sesionHttp)
    }

    private var cuerpoEnviado: [String: Any] {
        (try? JSONSerialization.jsonObject(with: buzon.ultimoCuerpo))
            as? [String: Any] ?? [:]
    }

    @Test("la barra final de la dirección no se cuela en las rutas")
    func barraFinal() async throws {
        let (sesionHttp, buzonPropio) = ServidorFalso.sesion()
        let conBarras = ClienteApi(urlBase: "https://api.ejemplo.com///", sesionHttp: sesionHttp)
        buzonPropio.respuesta = .json(
            #"{"tokenAcceso": "a", "expiraEn": 1, "tokenRefresco": "r"}"#
        )

        _ = try await conBarras.canjear(codigoCanje: "c")

        #expect(
            buzonPropio.ultimaPeticion?.url?.absoluteString
                == "https://api.ejemplo.com/auth/canjear"
        )
    }

    @Test("la dirección del login lleva los cuatro parámetros del contrato")
    func direccionDeLogin() throws {
        let url = try #require(
            cliente.urlIniciarLogin(
                proveedor: "google", estado: "abc123", esquema: "guardarenlaces")
        )
        let parametros = try #require(
            URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems)
        let porNombre = Dictionary(uniqueKeysWithValues: parametros.map { ($0.name, $0.value) })

        #expect(url.path == "/auth/iniciar")
        #expect(porNombre["proveedor"] == "google")
        #expect(porNombre["modo"] == "deeplink")
        #expect(porNombre["estado"] == "abc123")
        #expect(porNombre["esquema"] == "guardarenlaces")
    }

    @Test("canjear devuelve los tokens y con quién se ha entrado")
    func canjear() async throws {
        buzon.respuesta = .json(
            """
            {"tokenAcceso": "acceso", "expiraEn": 1735689600000, "tokenRefresco": "refresco",
             "usuario": {"id": 1, "email": "persona@ejemplo.com", "proveedor": "google"}}
            """
        )

        let respuesta = try await cliente.canjear(codigoCanje: "codigo")

        #expect(respuesta.tokenAcceso == "acceso")
        #expect(respuesta.tokenRefresco == "refresco")
        #expect(respuesta.usuario?.email == "persona@ejemplo.com")
        #expect(cuerpoEnviado["codigoCanje"] as? String == "codigo")
    }

    @Test("renovar sin usuario también vale: no todas las respuestas lo traen")
    func renovarSinUsuario() async throws {
        buzon.respuesta = .json(
            #"{"tokenAcceso": "a2", "expiraEn": 2, "tokenRefresco": "r2"}"#
        )

        let respuesta = try await cliente.renovar(tokenRefresco: "r1")

        #expect(respuesta.usuario == nil)
        #expect(respuesta.tokenRefresco == "r2")
    }

    @Test("las llamadas con sesión mandan el token en la cabecera")
    func cabeceraDeAutorizacion() async throws {
        buzon.respuesta = .json(
            #"{"titulo": "T", "descripcion": null, "imagenUrl": null, "tipo": "articulo"}"#
        )

        let metadatos = try await cliente.metadatos(url: "https://a.com", tokenAcceso: "elToken")

        #expect(
            buzon.ultimaPeticion?.value(forHTTPHeaderField: "Authorization")
                == "Bearer elToken"
        )
        #expect(metadatos.comoMetadatos.tipo == .articulo)
    }

    @Test("un 401 se reconoce como sesión caducada")
    func sesionCaducada() async throws {
        buzon.respuesta = .json(#"{"error": "token caducado"}"#, codigo: 401)

        await #expect(throws: ErrorApi.self) {
            _ = try await cliente.metadatos(url: "https://a.com", tokenAcceso: "viejo")
        }

        do {
            _ = try await cliente.metadatos(url: "https://a.com", tokenAcceso: "viejo")
        } catch let fallo as ErrorApi {
            #expect(fallo.esSesionCaducada)
            #expect(fallo.mensaje == "token caducado")
        }
    }

    @Test("sin conexión, el aviso está en castellano y no habla de sockets")
    func sinConexion() async throws {
        buzon.fallarLaConexion = true

        do {
            _ = try await cliente.canjear(codigoCanje: "c")
            Issue.record("tenía que haber fallado")
        } catch let fallo as ErrorApi {
            #expect(fallo.mensaje == "sin conexión con el servidor")
            #expect(fallo.codigo == nil)
        }
    }

    @Test("bajar cambios entiende enlaces y etiquetas en la misma respuesta")
    func bajarCambios() async throws {
        buzon.respuesta = .json(
            """
            {"elementos": [{"id": "e1", "url": "https://a.com", "actualizadoEn": 100}],
             "etiquetasDefinidas": [{"id": "t1", "nombre": "ocio", "actualizadoEn": 90}],
             "servidorEn": 1735600000123, "masDisponible": true}
            """
        )

        let respuesta = try await cliente.bajarCambios(desde: 50, tokenAcceso: "t")

        #expect(respuesta.elementos.map(\.id) == ["e1"])
        #expect(respuesta.etiquetasDefinidas.map(\.nombre) == ["ocio"])
        #expect(respuesta.servidorEn == 1_735_600_000_123)
        #expect(respuesta.masDisponible)
        #expect(buzon.ultimaPeticion?.url?.query?.contains("desde=50") == true)
    }

    @Test("una respuesta sin etiquetas no rompe nada")
    func respuestaSinEtiquetas() async throws {
        buzon.respuesta = .json(#"{"elementos": [], "servidorEn": 5}"#)

        let respuesta = try await cliente.bajarCambios(desde: 0, tokenAcceso: "t")

        #expect(respuesta.etiquetasDefinidas.isEmpty)
        #expect(respuesta.masDisponible == false)
    }

    @Test("al subir solo se manda lo que hay, no listas vacías")
    func subirSoloLoQueHay() async throws {
        buzon.respuesta = .json(#"{"elementos": [], "rechazados": []}"#)

        _ = try await cliente.subirCambios(
            elementos: [Elemento(id: "e1", url: "https://a.com", actualizadoEn: 100)],
            etiquetas: [],
            tokenAcceso: "t"
        )

        #expect(cuerpoEnviado["elementos"] != nil)
        #expect(cuerpoEnviado["etiquetasDefinidas"] == nil)
    }

    @Test("lo rechazado llega con su motivo, para poder sacarlo de la cola")
    func rechazados() async throws {
        buzon.respuesta = .json(
            """
            {"elementos": [], "rechazados": [{"id": "e1", "motivo": "no_aplicable"}],
             "etiquetasRechazadas": [{"id": "t1", "motivo": "sin_nombre"}]}
            """
        )

        let respuesta = try await cliente.subirCambios(
            elementos: [Elemento(id: "e1", url: "https://a.com")],
            etiquetas: [],
            tokenAcceso: "t"
        )

        #expect(respuesta.rechazados == [EntradaRechazada(id: "e1", motivo: "no_aplicable")])
        #expect(respuesta.etiquetasRechazadas.map(\.id) == ["t1"])
    }

    @Test("cerrar sesión aguanta que el servidor no conteste nada")
    func cerrarSesionSinCuerpo() async throws {
        buzon.respuesta = ServidorFalso.Respuesta(codigo: 200, cuerpo: Data())

        try await cliente.cerrarSesion(tokenRefresco: "r")

        #expect(buzon.ultimaPeticion?.url?.path == "/auth/logout")
    }
}

File: app-android/fontaneria/src/test/kotlin/com/jmortizsilva/guardarenlaces/fontaneria/PruebasClienteApi.kt
package com.jmortizsilva.guardarenlaces.fontaneria

import com.jmortizsilva.guardarenlaces.dominio.Elemento
import com.jmortizsilva.guardarenlaces.dominio.Proveedor
import com.jmortizsilva.guardarenlaces.dominio.TipoElemento
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.HttpUrl.Companion.toHttpUrl

class PruebasClienteApi {
    private val servidor = ServidorFalso()
    private val cliente = servidor.cliente()

    @AfterTest fun cerrar() = servidor.close()

    @Test
    fun `la barra final de la direccion no se cuela en las rutas`() = runBlocking {
        servidor.respuesta =
            ServidorFalso.Respuesta.json(
                """{"tokenAcceso": "a", "expiraEn": 1, "tokenRefresco": "r"}"""
            )

        servidor.cliente("https://api.ejemplo.com///").canjear("c")

        assertEquals(
            "https://api.ejemplo.com/auth/canjear",
            servidor.ultimaPeticion?.url.toString(),
        )
    }

    @Test
    fun `la direccion del login lleva los cuatro parametros del contrato`() {
        val url = cliente.urlIniciarLogin(Proveedor.Apple, estado = "abc123").toHttpUrl()

        assertEquals("/auth/iniciar", url.encodedPath)
        assertEquals("apple", url.queryParameter("proveedor"))
        assertEquals("deeplink", url.queryParameter("modo"))
        assertEquals("abc123", url.queryParameter("estado"))
        assertEquals("guardarenlaces", url.queryParameter("esquema"))
    }

    @Test
    fun `canjear devuelve los tokens y con quien se ha entrado`() = runBlocking {
        servidor.respuesta =
            ServidorFalso.Respuesta.json(
                """{"tokenAcceso": "acceso", "expiraEn": 1735689600000, "tokenRefresco": "refresco",
                   "usuario": {"id": 1, "email": "persona@ejemplo.com", "proveedor": "google"}}"""
            )

        val respuesta = cliente.canjear("codigo")

        assertEquals("acceso", respuesta.tokenAcceso)
        assertEquals("refresco", respuesta.tokenRefresco)
        assertEquals("persona@ejemplo.com", respuesta.usuario?.email)
        assertEquals(
            "codigo",
            servidor.ultimaPeticion?.cuerpoJson?.get("codigoCanje")?.jsonPrimitive?.content,
        )
    }

    @Test
    fun `renovar sin usuario tambien vale, no todas las respuestas lo traen`() = runBlocking {
        servidor.respuesta =
            ServidorFalso.Respuesta.json(
                """{"tokenAcceso": "a2", "expiraEn": 2, "tokenRefresco": "r2"}"""
            )

        val respuesta = cliente.renovar("r1")

        assertNull(respuesta.usuario)
        assertEquals("r2", respuesta.tokenRefresco)
    }

    @Test
    fun `las llamadas con sesion mandan el token en la cabecera`() = runBlocking {
        servidor.respuesta =
            ServidorFalso.Respuesta.json(
                """{"titulo": "T", "descripcion": null, "imagenUrl": null, "tipo": "articulo"}"""
            )

        val metadatos = cliente.metadatos("https://a.com", "elToken")

        assertEquals("Bearer elToken", servidor.ultimaPeticion?.cabeceras?.get("Authorization"))
        assertEquals(TipoElemento.Articulo, metadatos.comoMetadatos.tipo)
    }

    @Test
    fun `un 401 se reconoce como sesion caducada`() = runBlocking {
        servidor.respuesta = ServidorFalso.Respuesta.json("""{"error": "token caducado"}""", 401)

        val fallo = assertFailsWith<ErrorApi> { cliente.metadatos("https://a.com", "viejo") }

        assertTrue(fallo.esSesionCaducada)
        assertEquals("token caducado", fallo.mensaje)
    }

    @Test
    fun `un error sin cuerpo del contrato dice el codigo`() = runBlocking {
        servidor.respuesta = ServidorFalso.Respuesta(502, "<html>Bad Gateway</html>")

        val fallo = assertFailsWith<ErrorApi> { cliente.metadatos("https://a.com", "t") }

        assertEquals("el servidor respondió con un error 502", fallo.mensaje)
    }

    @Test
    fun `sin conexion, el aviso esta en castellano y no habla de sockets`() = runBlocking {
        servidor.fallarLaConexion = true

        val fallo = assertFailsWith<ErrorApi> { cliente.canjear("c") }

        assertEquals("sin conexión con el servidor", fallo.mensaje)
        assertNull(fallo.codigo)
    }

    @Test
    fun `una respuesta que no se entiende se dice, no revienta`() = runBlocking {
        servidor.respuesta = ServidorFalso.Respuesta.json("esto no es json")

        val fallo = assertFailsWith<ErrorApi> { cliente.canjear("c") }

        assertEquals("el servidor respondió algo que no se entiende", fallo.mensaje)
    }

    @Test
    fun `bajar cambios entiende enlaces y etiquetas en la misma respuesta`() = runBlocking {
        servidor.respuesta =
            ServidorFalso.Respuesta.json(
                """{"elementos": [{"id": "e1", "url": "https://a.com", "actualizadoEn": 100}],
                   "etiquetasDefinidas": [{"id": "t1", "nombre": "ocio", "actualizadoEn": 90}],
                   "servidorEn": 1735600000123, "masDisponible": true}"""
            )

        val respuesta = cliente.bajarCambios(desde = 50, tokenAcceso = "t")

        assertEquals(listOf("e1"), respuesta.elementos.map { it.id })
        assertEquals(listOf("ocio"), respuesta.etiquetasDefinidas.map { it.nombre })
        assertEquals(1_735_600_000_123, respuesta.servidorEn)
        assertTrue(respuesta.masDisponible)
        assertEquals("50", servidor.ultimaPeticion?.url?.queryParameter("desde"))
        assertEquals("GET", servidor.ultimaPeticion?.metodo)
    }

    @Test
    fun `una respuesta sin etiquetas no rompe nada`() = runBlocking {
        servidor.respuesta = ServidorFalso.Respuesta.json("""{"elementos": [], "servidorEn": 5}""")

        val respuesta = cliente.bajarCambios(desde = 0, tokenAcceso = "t")

        assertTrue(respuesta.etiquetasDefinidas.isEmpty())
        assertFalse(respuesta.masDisponible)
    }

    @Test
    fun `al subir solo se manda lo que hay, no listas vacias`() = runBlocking {
        servidor.respuesta = ServidorFalso.Respuesta.json("""{"elementos": [], "rechazados": []}""")

        cliente.subirCambios(
            elementos = listOf(Elemento(id = "e1", url = "https://a.com", actualizadoEn = 100)),
            etiquetas = emptyList(),
            tokenAcceso = "t",
        )

        val cuerpo = servidor.ultimaPeticion!!.cuerpoJson
        assertTrue("elementos" in cuerpo)
        assertFalse("etiquetasDefinidas" in cuerpo)
    }

    @Test
    fun `lo rechazado llega con su motivo, para poder sacarlo de la cola`() = runBlocking {
        servidor.respuesta =
            ServidorFalso.Respuesta.json(
                """{"elementos": [], "rechazados": [{"id": "e1", "motivo": "no_aplicable"}],
                   "etiquetasRechazadas": [{"id": "t1", "motivo": "sin_nombre"}]}"""
            )

        val respuesta =
            cliente.subirCambios(
                listOf(Elemento(id = "e1", url = "https://a.com")),
                emptyList(),
                "t",
            )

        assertEquals(listOf(EntradaRechazada("e1", "no_aplicable")), respuesta.rechazados)
        assertEquals(listOf("t1"), respuesta.etiquetasRechazadas.map { it.id })
    }

    @Test
    fun `cerrar sesion aguanta que el servidor no conteste nada`() = runBlocking {
        servidor.respuesta = ServidorFalso.Respuesta(200, "")

        cliente.cerrarSesion("r")

        assertEquals("/auth/logout", servidor.ultimaPeticion?.ruta)
    }
}
